relay-server/src/actors/project_cache.rs

```diff
@@ -242,69 +231,218 @@ impl Handler<FetchProjectState> for ProjectCache {
+                match relay_mode {
+                    RelayMode::Proxy => {
+                        return Box::new(future::ok(ProjectStateResponse::new(Arc::new(
+                            ProjectState::allowed(),
+                        ))));
+                    }
+                    RelayMode::Static => {
+                        return Box::new(future::ok(ProjectStateResponse::new(Arc::new(
+                            ProjectState::missing(),
+                        ))));
+                    }
+                    RelayMode::Capture => {
+                        return Box::new(future::ok(ProjectStateResponse::new(Arc::new(
+                            ProjectState::allowed(),
+                        ))));
+                    }
+                    RelayMode::Managed => {
+                        // Proceed with loading the config from redis or upstream
+                    }
+                }
 
-                Box::new(fetch_upstream)
-            });
+                #[cfg(not(feature = "processing"))]
+                let fetch_redis = future::ok(None);
+
+                #[cfg(feature = "processing")]
+                let fetch_redis: ResponseFuture<_, _> = if let Some(ref redis_source) = redis_source
+                {
+                    Box::new(
+                        redis_source
+                            .send(FetchOptionalProjectState { public_key })
+                            .map_err(|_| ()),
+                    )
+                } else {
+                    Box::new(future::ok(None))
+                };
+
+                let fetch_redis = fetch_redis.and_then(move |response| {
+                    if let Some(state) = response {
+                        return Box::new(future::ok(ProjectStateResponse::new(state)))
+                            as ResponseFuture<_, _>;
+                    }
+
+                    let fetch_upstream = upstream_source
+                        .send(FetchProjectState {
+                            public_key,
+                            no_cache,
+                        })
+                        .map_err(|_| ())
+                        .and_then(move |result| result.map_err(|_| ()));
+
+                    Box::new(fetch_upstream)
+                });
 
-            Box::new(fetch_redis)
-        });
+                Box::new(fetch_redis)
+            })
+            .into_actor(self)
+            .then(move |state_result, slf, context| {
+                let proj = slf.get_project(public_key);
+                proj.update_state(state_result, no_cache, context);
+                fut::ok::<_, (), _>(())
+            })
+            .spawn(context);
+    }
+}
+
+impl Handler<GetProjectState> for ProjectCache {
+    type Result = Response<Arc<ProjectState>, ProjectError>;
+
+    fn handle(&mut self, message: GetProjectState, context: &mut Context<Self>) -> Self::Result {
+        let proj = self.get_project(message.public_key);
```

Comment:
Also applies to all other instances: 

```suggestion
        let project = self.get_project(message.public_key);
```